File: Parallel Programming - Weeks 4-7/Week 5/1 - Hello Threads/HelloThread.java
/*
 * The smallest useful Java threaded program. One background
 * thread says hello, main prints its own line, and then main
 * waits for the background thread to finish before exiting.
 *
 * Compare with hello_thread.c in the same folder. Run them both
 * and you should see the same output, up to ordering.
 */
public class HelloThread {

    public static void main(String[] args) throws InterruptedException {

        // A thread built from a lambda. The lambda is the "run"
        // method of an anonymous Runnable.
        Thread t = new Thread(() -> {
            System.out.println("Hello from the child thread!");
        });

        // Java needs you to explicitly start the thread. Calling
        // run() directly would just execute the lambda on the
        // main thread, which is a classic beginner mistake.
        t.start();

        // Runs in main, at the same time as the child thread. The
        // order in which the two lines print is up to the OS.
        System.out.println("Hello from the main thread!");

        // Wait for the child to finish before we exit.
        t.join();

        System.out.println("Both threads are done. Goodbye!");
    }
}

File: Parallel Programming - Weeks 4-7/Week 5/1 - Hello Threads/hello_thread.c
/*
 * Same program as HelloThread.java, rewritten with pthreads.
 *
 * A few things worth noticing next to the Java version:
 *
 *   - The thread body has to be a function with the exact shape
 *         void* fn(void* arg)
 *     There are no lambdas and no anonymous Runnables.
 *
 *   - There is no separate "start" step. pthread_create both
 *     creates AND starts the thread in a single call.
 *
 *   - pthread_join takes a pointer for the thread's return value.
 *     We don't care about a return value here, so we pass NULL.
 */

#include <stdio.h>
#include <pthread.h>

/* The thread body. */
void* say_hello(void* arg) {
    printf("Hello from the child thread!\n");
    return NULL;
}

int main() {
    pthread_t t;

    /* Create AND start the thread in one call. No .start() step. */
    pthread_create(&t, NULL, say_hello, NULL);

    printf("Hello from the main thread!\n");

    /* Wait for the child thread to finish. Same role as t.join() in Java. */
    pthread_join(t, NULL);

    printf("Both threads are done. Goodbye!\n");
    return 0;
}
